snakemake/executors.py

```diff
@@ -2115,3 +2128,759 @@ def run_wrapper(
+
+        # https://cloud.google.com/life-sciences/docs/reference/rest/v2beta/projects.locations.pipelines
+        pipelines = self._api.projects().locations().pipelines()
+
+        # pipelines.run
+        # https://cloud.google.com/life-sciences/docs/reference/rest/v2beta/projects.locations.pipelines/run
+
+        labels = self._generate_pipeline_labels(job)
+        pipeline = self._generate_pipeline(job)
+
+        # The body of the request is a Pipeline and labels
+        body = {"pipeline": pipeline, "labels": labels}
+
+        # capabilities - this won't currently work (Singularity in Docker)
+        # We either need to add CAPS or run in privileged mode (ehh)
+        if job.needs_singularity and self.workflow.use_singularity:
+            logger.warning(
+                "Singularity requires additional capabilities that "
+                "aren't yet supported for standard Docker runs."
+            )
+
+        # location looks like: "projects/<project>/locations/<location>"
+        operation = pipelines.run(parent=self.location, body=body)
+
+        # 403 will result if no permission to use pipelines or project
+        result = self._retry_request(operation)
+
+        # The jobid is the last number of the full name
+        jobid = result["name"].split("/")[-1]
+
+        # Give some logging for how to get status
+        logger.info(
+            "Get status with:\n"
+            "gcloud config set project {project}\n"
+            "gcloud beta lifesciences operations describe {jobid}\n"
+            "gcloud beta lifesciences operations list".format(
+                project=self.project, jobid=jobid
+            )
+        )
+
+        self.active_jobs.append(
+            GoogleLifeSciencesJob(job, result["name"], jobid, callback, error_callback)
+        )
+
+    def _job_was_successful(self, status):
+        """based on a status response (a [pipeline].projects.locations.operations.get
+           debug print the list of events, return True if all return codes 0
+           and False otherwise (indication of failure). In that a nonzero exit
+           status is found, we also debug print it for the user.
+        """
+        success = True
+
+        # https://cloud.google.com/life-sciences/docs/reference/rest/v2beta/Event
+        for event in status["metadata"]["events"]:
+
+            logger.info(event["description"])
+
+            # Does it always result in fail for other failure reasons?
+            if "failed" in event:
+                success = False
+                action = event.get("failed")
+                logger.error("{}: {}".format(action["code"], action["cause"]))
+
+            elif "unexpectedExitStatus" in event:
+                action = event.get("unexpectedExitStatus")
+
+                if action["exitStatus"] != 0:
+                    success = False
+
+                    # Provide reason for the failure (desc includes exit code)
+                    msg = "%s" % event["description"]
```

Comment:
The above line seems to be superfluous to me.

Reply:
Ah, do you make a copy of the string in the event?

Reply:
I can definitely remove it - I was thinking we'd want to print to the console and to whatever error is defined. Would you like me to remove?

Reply:
The goal of adding the event description to the error message was to give the user that reads the error log context about the action step that the error output is associated with.
```
                    msg = "%s" % event["description"]
                    if "stderr" in action:
                        msg += ": %s" % action["stderr"]
                        logger.error(msg)
```
I can remove it if you think redundant, but then the logger error won't have the description of the event that failed.

Reply:
You can keep the description I think, but let us make this debug for now. Usually the message that snakemake prints in the master process (the `error in rule ...` thingy) is enough. Having a huge output on error just scares people and distracts them from the true information (which usually is contained in the log file of the rule and not in global stderr or stdout).

Reply:
Gotcha! This is now fixed, and pushed. I'm going to update the branches I have running the test pipeline and I'll show you the updated output in the comment below.